timeout_handler.go

```diff
@@ -0,0 +1,231 @@
+	}
+}
+
+func (t *TimeoutHandler) Tick(now time.Time) {
+	select {
+	case t.ticks <- now:
+	default:
+		t.log.Debug("Dropping tick in timeouthandler")
+	}
+}
+
+func (t *TimeoutHandler) AddTask(task *TimeoutTask) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	if _, ok := t.tasks[string(task.NodeID)]; !ok {
+		t.log.Info("Attempting to add a task for an unknown node", zap.Stringer("node", task.NodeID))
+	}
+
+	// adds a task to the heap and the tasks map
+	if _, ok := t.tasks[string(task.NodeID)][task.TaskID]; ok {
+		t.log.Debug("Trying to add an already included task", zap.Stringer("node", task.NodeID), zap.String("Task ID", task.TaskID))
+		return
+	}
+
+	t.tasks[string(task.NodeID)][task.TaskID] = task
+	t.log.Debug("Adding timeout task", zap.Stringer("node", task.NodeID), zap.String("taskid", task.TaskID))
+	heap.Push(&t.heap, task)
+}
+
+func (t *TimeoutHandler) RemoveTask(nodeID NodeID, ID string) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	if _, ok := t.tasks[string(nodeID)]; !ok {
+		t.log.Info("Attempting to add a remove a task for an unknown node", zap.Stringer("node", nodeID))
+		return
+	}
+
+	if _, ok := t.tasks[string(nodeID)][ID]; !ok {
+		return
+	}
+
+	// find the task using the task map
+	// remove it from the heap using the index
+	t.log.Debug("Removing timeout task", zap.String("taskid", ID))
+	heap.Remove(&t.heap, t.tasks[string(nodeID)][ID].index)
+	delete(t.tasks[string(nodeID)], ID)
+}
+
+func (t *TimeoutHandler) Close() {
+	select {
+	case <-t.close:
+		return
+	default:
+		close(t.close)
+	}
+}
+
+// FindTask returns the first TimeoutTask assigned to [node] that contains any sequence in [seqs].
+// A sequence is considered "contained" if it falls between a task's Start (inclusive) and End (inclusive).
+func (t *TimeoutHandler) FindTask(node NodeID, seqs []uint64) *TimeoutTask {
+	var task *TimeoutTask
+seq:
+	for _, seq := range seqs {
+		for _, t := range t.tasks[string(node)] {
+			if seq >= t.Start && seq <= t.End {
+				task = t
+				break seq
+			}
+		}
+	}
+
+	return task
+}
+
+const delimiter = "_"
+
+func getTimeoutID(start, end uint64) string {
+	return fmt.Sprintf("%d%s%d", start, delimiter, end)
+}
+
+// ----------------------------------------------------------------------
+type TaskHeap []*TimeoutTask
+
+func (h *TaskHeap) Len() int { return len(*h) }
+
+// Less returns if the task at index [i] has a lower timeout than the task at index [j]
+func (h *TaskHeap) Less(i, j int) bool { return (*h)[i].Deadline.Before((*h)[j].Deadline) }
+
+// Swap swaps the values at index [i] and [j]
+func (h *TaskHeap) Swap(i, j int) {
+	(*h)[i], (*h)[j] = (*h)[j], (*h)[i]
+	(*h)[i].index = i
+	(*h)[j].index = j
+}
+
+func (h *TaskHeap) Push(x any) {
+	task := x.(*TimeoutTask)
+	task.index = h.Len()
+	*h = append(*h, task)
+}
+
+func (h *TaskHeap) Pop() any {
+	old := *h
+	len := h.Len()
+	task := old[len-1]
+	old[len-1] = nil
+	*h = old[0 : len-1]
+	task.index = -1
+	return task
+}
+
+func (h *TaskHeap) Peep() any {
```

Comment:
why do we need `Peep()` ? I don't think it's used anywhere. 

To implement a head we just need:

```
type Interface interface {
	sort.Interface
	Push(x any) // add x as element Len()
	Pop() any   // remove and return element Len() - 1.
}
```

Also, did you mean `Peek()` ?